healvis/visibility.py

```diff
@@ -46,79 +46,211 @@ def jy2Tstr(f, bm=1.0):
+            sigma : float, standard deviation [degrees] for gaussian beam
+            diameter : float, dish diameter [meter] used for airy beam
+
+        Notes:
+            Uniform beam is a flat-top beam across the entire sky.
+            Gaussian beam is a frequency independent, peak normalized Gaussian
+                with respect to zenith angle.
+            Airy beam uses the dish diameter to set the airy beam width as a function of frequency.
+            callable is any function that takes (za_array, freqs) in units [radians, Hz] respectively
+                and returns an ndarray of shape (Nfreqs, Nza_array) with power beam values.
+        """
+        if beam_type not in ['uniform', 'gaussian', 'airy'] and not callable(beam_type):
             raise NotImplementedError("Beam type " + str(beam_type) + " not available yet.")
         self.beam_type = beam_type
         if beam_type == 'gaussian':
             if sigma is None:
                 raise KeyError("Sigma required for gaussian beam")
             self.sigma = sigma * np.pi / 180.  # deg -> radians
-        if beam_type == 'airy':
+        elif beam_type == 'airy':
             if diameter is None:
                 raise KeyError("Dish diameter required for airy beam")
             self.diameter = diameter
 
-    def plot_beam(self, az, za):
+    def plot_beam(self, az, za, freqs, **kwargs):
+        ## TODO: needs development, and testing coverage?
         fig = pl.figure()
-        pl.imshow(self.beam_val(az, za))
+        pl.imshow(self.beam_val(az, za, freqs, **kwargs))
         pl.show()
 
-    def beam_val(self, az, za, freq_Hz):
+    def beam_val(self, az, za, freqs, **kwargs):
         """
-        az, za = radians
+        Evaluation of an analytic beam model.
+
+        Args:
+            az : float or ndarray, azimuth angle [radian], must have len(za)
+            za : float or ndarray, zenith angle [radian], must have len(az)
+            freqs : float or ndarray, frequencies [Hz]
+            kwargs : keyword arguments to pass if self.beam_type is callable
 
+        Returns:
+            beam_value : ndarray of beam power, with shape (Nfreqs, Naz_array)
         """
         if self.beam_type == 'uniform':
             if isinstance(az, np.ndarray):
-                return np.ones_like(az)
-            return 1
-        if self.beam_type == 'gaussian':
-            return np.exp(-(za**2) / (2 * self.sigma**2))  # Peak normalized
+                beam_value = np.ones((len(freqs), len(az)), dtype=np.float)
+            else:
+                beam_value = 1.0
+        elif self.beam_type == 'gaussian':
+            beam_value = np.exp(-(za**2) / (2 * self.sigma**2))  # Peak normalized
+            beam_value = np.repeat(beam_value[None], len(freqs), axis=0)
+        elif self.beam_type == 'airy':
+            beam_value = airy_disk(za, freqs, diameter=self.diameter)
```

Comment:
Ah, because it seemed simple enough I guess. I can make it its own function if you like, but the airy disk was a little more complicated so I just made it its own function for modularity. I'm fine w/ things as is if you are.

Reply:
That makes sense. I see also that you used this function separately in a test, so it makes sense to pull it out.